flax/nnx/transforms/transforms.py

```diff
@@ -160,3 +164,141 @@ def cond(
     *operands,
     **kwargs,
   )
+
+
+@general.split_inputs(ctxtag='switch')
+def switch(
+  index,
+  branches: tp.Sequence[tp.Callable[..., A]],
+  *operands,
+) -> A:
+  return jax.lax.switch(
+    index,
+    [general.merge_inputs(f, ctxtag='switch') for f in branches],
+    *operands,
+  )
+
+
+# -------------------------------
+# while_loop
+# -------------------------------
+
+
+@dataclasses.dataclass(eq=False)
+class WhileLoopCondFn:
+  f: tp.Callable[..., tp.Any]
+
+  def __post_init__(self):
+    functools.update_wrapper(self, self.f)
+
+  def __call__(self, pure_val):
+    val = extract.from_tree(pure_val)
+    out = self.f(val)
+    return out
+
+
+def _add_fake_index_mapping(tree: tp.Any):
+  def per_node_state(ns: extract.NodeStates | tp.Any):
+    global_index_mapping = {}
+    if not isinstance(ns, extract.NodeStates):
+      return ns
+    assert isinstance(ns._graphdef, graph.NodeDef)
```

Comment:
When you have a repeated NNX object in the pytree processed by `to_tree` you will get a `NodeRef` here. You can just return in that case. Fix:

```suggestion
    if not isinstance(ns._graphdef, graph.NodeDef):
      return ns
```
I tested this example that fails in this branch.
```python
m = nnx.Linear(10, 10, rngs=nnx.Rngs(0))

def body_fn(val):
  count, m, _ = val
  return count + 1, m, m

count, m, _ = nnx.while_loop(
  lambda val: val[0] < 2,
  body_fn,
  (0, m, m),
)
```